File: go/kata/game.go
package kata

type Game struct {
	currentRoll int
	rolls [21]int
}

func(g* Game) roll(pins int) {
	g.rolls[g.currentRoll] = pins
	if pins == 10 && g.currentRoll % 2 == 0 {
		g.currentRoll += 2
	} else {
		g.currentRoll += 1
	}
}

func(g* Game) score() int {
	score := 0
	for frame := 0; frame < 10; frame++ {
		roll := g.rolls[frame*2] + g.rolls[frame*2+1]
		score += roll
		if isStrike(g, frame) {
			score +=  g.rolls[frame*2+2] + g.rolls[frame*2+3]
		} else if isSpare(g, frame) {
			score += g.rolls[frame*2+2]
		}
	}
	return score
}

func isStrike(g* Game, frame int)  bool {
	return g.rolls[frame*2] == 10
}

func isSpare(g* Game, frame int) bool {
	return g.rolls[frame*2] + g.rolls[frame*2] == 10 &&
		g.rolls[frame*2] != 10
}

File: ruby/kata/game.rb
class Game

  def initialize
    @current_roll = 0
    @rolls = []
  end

  def roll(pins)
    @rolls[@current_roll] = pins
    @current_roll += 1
  end

  def score
    score = 0
    for frame in (0..19).step(2) do  
      if is_spare?(frame)
        score += 10 + @rolls[frame+2]
      else
        score += @rolls[frame] + @rolls[frame+1]
      end
    end
    return score
  end

  def is_spare?(frame)
    return @rolls[frame] + @rolls[frame+1] == 10
  end
end
